File: src/Args.cpp
/*
** EPITECH PROJECT, 2023
** Crypto
** File description:
** Args
*/

#include "Args.hpp"

#include <iostream>
#include <fstream>
#include <stdexcept>
#include <vector>

namespace PGP
{

Args::Args()
{
    _mode = Mode::NONE;
    _algorithm = Algorithm::NONE;
    _help = false;
    _block = false;
    _p = 0;
    _q = 0;
    _key = "";
}

Args::~Args()
{

}

void Args::parse(int argc, char* argv[])
{
    std::vector<std::string> args;

    for (int i = 1; i < argc; i++) {
        args.push_back(argv[i]);
    }

    for (auto it = args.begin(); it != args.end(); it++) {
        std::string arg = *it;
        if (arg == "-h" || arg == "--help") {
            _help = true;
            return;
        }
        if (arg == "-xor" && _algorithm == Algorithm::Algorithm::NONE) {
            _algorithm = Algorithm::XOR;
            it = args.erase(it).operator--();
        }
        if (arg == "-rsa" && _algorithm == Algorithm::Algorithm::NONE) {
            _algorithm = Algorithm::RSA;
            it = args.erase(it).operator--();
        }
        if (arg == "-aes" && _algorithm == Algorithm::Algorithm::NONE) {
            _algorithm = Algorithm::AES;
            it = args.erase(it).operator--();
        }
        if (arg == "-c" && _mode == Mode::Mode::NONE) {
            _mode = Mode::ENCRYPT;
            it = args.erase(it).operator--();
        }
        if (arg == "-d" && _mode == Mode::Mode::NONE) {
            _mode = Mode::DECRYPT;
            it = args.erase(it).operator--();
        }
        if (arg == "-b" && !_block) {
            _block = true;
            it = args.erase(it).operator--();
        }
        if (arg == "-g" && _mode == Mode::Mode::NONE) {
            _mode = Mode::GEN_RSA_KEY;
            it = args.erase(it).operator--();
        }
    }
    if (_help)
        return;
    if (_mode == Mode::NONE || _algorithm == Algorithm::NONE)
        throw std::runtime_error("Invalid arguments");
    if (_block && (_algorithm != Algorithm::XOR && _algorithm != Algorithm::AES))
        throw std::runtime_error("Invalid arguments");
    if (args.empty()) {
        throw std::runtime_error("Invalid arguments");
    }
    if (_mode == Mode::GEN_RSA_KEY && _algorithm != Algorithm::RSA) {
        throw std::runtime_error("Invalid arguments");
    }
    if (_mode == Mode::GEN_RSA_KEY) {
        if (args.size() < 2)
            throw std::runtime_error("Invalid arguments");
        return;
    }
    _key = args.front();
}

void Args::printHelp() const
{
    std::ifstream file(PATH_HELP);
    std::string buffer;

    while (std::getline(file, buffer)) {
        std::cout << buffer << std::endl;
    }
}

} // namespace PGP

File: src/args.py
# args.py

import sys
from enum import Enum

class Mode(Enum):
    NONE = -1,
    ENCRYPT = 0,
    DECRYPT = 1,
    GEN_RSA_KEY = 2,

class Algorithm(Enum):
    NONE = -1
    XOR = 0
    RSA = 1
    AES = 2
    PGP = 3
    SIZE = 4

PATH_HELP = "assets/help.txt"

class Args:
    def __init__(self):
        self.mode = Mode.NONE
        self.algo = Algorithm.NONE
        self.help = False
        self.block = False
        self.P = 0
        self.Q = 0
        self.key = ""

    def parse(self):

        list = sys.argv
        del list[0]
        if (not list):
            raise("no arguments provided")

        for arg in list[:]:
            if (arg == "-h" or arg == "--help"):
                self.help = True
                return
            if (arg == "-xor"):
                self.algo = Algorithm.XOR
                list.remove(arg)
            if (arg == "-rsa"):
                self.algo = Algorithm.RSA
                list.remove(arg)
            if (arg == "-aes"):
                self.algo = Algorithm.AES
                list.remove(arg)
            if (arg == "-pgp"):
                self.algo = Algorithm.PGP
                list.remove(arg)
            if (arg == "-c"):
                self.mode = Mode.ENCRYPT
                list.remove(arg)
            if (arg == "-d"):
                self.mode = Mode.DECRYPT
                list.remove(arg)
            if (arg == "-b"):
                self.block = True
                list.remove(arg)
            if (arg == "-g"):
                self.mode = Mode.GEN_RSA_KEY
                list.remove(arg)

        if (self.mode == Mode.NONE):
            raise ValueError("can't find encrypt or decrypt mode")
        if (self.algo == Algorithm.NONE):
            raise ValueError("can't find algorithm to use")
        if (self.block and (self.algo != Algorithm.XOR and self.algo != Algorithm.AES)):
            raise ValueError("can't find algorithm to use")
        if (self.mode == Mode.GEN_RSA_KEY and self.algo != Algorithm.RSA):
            raise ValueError("-g option must be with rsa algorithm")
        if (self.mode == Mode.GEN_RSA_KEY):
            if (len(list) < 2):
                raise ValueError("not enough arguments")
            self.P = int.from_bytes(bytes.fromhex(list[0]), "little")
            self.Q = int.from_bytes(bytes.fromhex(list[1]), "little")
            return
        self.key = list[0]

    def printHelp(self):
        print(open(PATH_HELP, 'r').read())
